File: ga/4omini_result_2024_7/result/mutex_mini/8_mutex.c
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

fn thread1_function(mutex: Arc<Mutex<()>>) {
    let loop_count = 2;

    for _ in 0..loop_count {
        // Lock the mutex for mutual exclusion
        let _lock = mutex.lock().unwrap();
        
        for _ in 0..10 {
            println!("Hello world");
        }
        
        // Mutex is automatically unlocked when going out of scope
        thread::sleep(Duration::from_secs(1));
    }
}

fn thread2_function(mutex: Arc<Mutex<()>>) {
    let loop_count = 2;

    thread::sleep(Duration::from_secs(1)); // Sleep before starting

    for _ in 0..loop_count {
        // Lock the mutex for mutual exclusion
        let _lock = mutex.lock().unwrap();
        
        for _ in 0..10 {
            println!("Good morning");
        }
        
        // Mutex is automatically unlocked when going out of scope
        thread::sleep(Duration::from_secs(1));
    }
}

fn main() {
    // Create a mutex and wrap it in an Arc for shared ownership between threads
    let mutex = Arc::new(Mutex::new(()));

    // Clone the Arc to pass to threads
    let mutex_clone1 = Arc::clone(&mutex);
    let mutex_clone2 = Arc::clone(&mutex);

    // Create the threads
    let thread1 = thread::spawn(move || thread1_function(mutex_clone1));
    let thread2 = thread::spawn(move || thread2_function(mutex_clone2));

    // Wait for both threads to finish
    thread1.join().unwrap();
    thread2.join().unwrap();

    println!("The thread is over, process is over too.");
}

File: ga/glm4_result_2024_6/result/mutex_glm/8_mutex.rs
use std::thread;
use std::sync::{Mutex, Arc};
use std::time::Duration;

fn main() {
    let mut threads = Vec::new();
    let counter = Arc::new(Mutex::new(0));

    for _ in 0..2 {
        let counter = Arc::clone(&counter);
        let handle = thread::spawn(move || {
            let mut loop_count = 0;
            while loop_count < 2 {
                let mut counter = counter.lock().unwrap();
                for _ in 0..10 {
                    println!("Hello world");
                }
                drop(counter);
                thread::sleep(Duration::from_secs(1));
                loop_count += 1;
            }
        });
        threads.push(handle);
    }

    for thread in threads {
        thread.join().unwrap();
    }

    println!("The threads are over, process is over too.");
}

fn thread1_function(counter: Arc<Mutex<i32>>) {
    let mut loop_count = 0;
    while loop_count < 2 {
        let mut counter = counter.lock().unwrap();
        for _ in 0..10 {
            println!("Hello world");
        }
        drop(counter);
        thread::sleep(Duration::from_secs(1));
        loop_count += 1;
    }
}

fn thread2_function(counter: Arc<Mutex<i32>>) {
    let mut loop_count = 0;
    while loop_count < 2 {
        let mut counter = counter.lock().unwrap();
        for _ in 0..10 {
            println!("Good morning");
        }
        drop(counter);
        thread::sleep(Duration::from_secs(1));
        loop_count += 1;
    }
}